components/brave_wallet/browser/simple_hash_client.cc

```diff
@@ -0,0 +1,639 @@
+  //           }
+  //         ],
+  //         "is_normalized": true,
+  //         "name_length": 16,
+  //         "segment_length": 16,
+  //         "version": 0,
+  //         "background_image":
+  //         "https://metadata.ens.domains/mainnet/avatar/stochasticparrot.eth",
+  //         "image_url":
+  //         "https://metadata.ens.domains/mainnet/0x57f1887a8BF19b14fC0dF6Fd9B2acc9Af147eA85/0x76dc36f2ff546436694c7eee18598a1309f0d382934ac4fd977ed24f3b9bb6a0/image",
+  //         "image_original_url":
+  //         "https://metadata.ens.domains/mainnet/0x57f1887a8BF19b14fC0dF6Fd9B2acc9Af147eA85/0x76dc36f2ff546436694c7eee18598a1309f0d382934ac4fd977ed24f3b9bb6a0/image",
+  //         "animation_original_url": null,
+  //         "metadata_original_url":
+  //         "https://metadata.ens.domains/mainnet/0x57f1887a8BF19b14fC0dF6Fd9B2acc9Af147eA85/53762001732575849527995595036249427730510390651723189221519398504820492711584/"
+  //       }
+  //     },
+  //     ...
+  // }
+
+  // Only ETH and SOL NFTs are supported.
+  if (!(coin == mojom::CoinType::ETH || coin == mojom::CoinType::SOL)) {
+    return absl::nullopt;
+  }
+
+  // If both skip_spam and only_spam are true, return early.
+  if (skip_spam && only_spam) {
+    return absl::nullopt;
+  }
+
+  const base::Value::Dict* dict = json_value.GetIfDict();
+  if (!dict) {
+    return absl::nullopt;
+  }
+
+  auto* next_cursor_ptr = dict->FindString("next_cursor");
+  absl::optional<std::string> next_cursor;
+  if (next_cursor_ptr) {
+    next_cursor = *next_cursor_ptr;
+  } else {
+    next_cursor = absl::nullopt;
+  }
+
+  const base::Value::List* nfts = dict->FindList("nfts");
+  if (!nfts) {
+    return absl::nullopt;
+  }
+
+  std::vector<mojom::BlockchainTokenPtr> nft_tokens;
+  for (const auto& nft_value : *nfts) {
+    auto token = mojom::BlockchainToken::New();
+
+    const base::Value::Dict* nft = nft_value.GetIfDict();
+    if (!nft) {
+      continue;
+    }
+    auto* collection = nft->FindDict("collection");
+    if (!collection) {
+      continue;
+    }
+    absl::optional<int> spam_score = collection->FindInt("spam_score");
+    if (skip_spam && (!spam_score || *spam_score > 0)) {
+      continue;
+    }
+    if (only_spam && (spam_score && *spam_score <= 0)) {
+      continue;
+    }
+
+    // contract_address (required)
+    auto* contract_address = nft->FindString("contract_address");
+    if (!contract_address) {
+      continue;
+    }
+    token->contract_address = *contract_address;
+
+    // chain_id (required)
+    auto* chain = nft->FindString("chain");
+    if (!chain) {
+      continue;
+    }
+    absl::optional<std::string> chain_id = SimpleHashChainIdToChainId(*chain);
+    if (!chain_id) {
+      continue;
+    }
+    token->chain_id = *chain_id;
+
+    // name
+    auto* name = nft->FindString("name");
+    if (name) {
+      token->name = *name;
+    }
+
+    // logo
+    auto* logo = nft->FindString("image_url");
```

Comment:
Can we fetch the `image_url` if the NFT is not considered spam and use the cdn url if it is considered spam? The main reason for this would be to make it so that spam NFT images can't be used as a mechanism to detect the users IP and because simplehash will do a bit of pre-image processing to convert it to a PNG which is less likely to be able to behave in unexpected ways.

Reply:
I believe the `image_url` referenced here points to SimpleHash's cdn URL, e.g.  https://cdn.simplehash.com/assets/0b270f746cda7694c9d6dad5d69384647092cc9016702110fdf2651b0f052641.jpg.  The SimpleHash response also includes a `image_original_url` field which points to the image URL included in the metadata, but we're not using it.  I believe that addresses your point on user IP addresses.

Reply:
Yup, that would address it thank you for pointing out that distinction. I got confused because IIRC `image_url` is the property name from the 721 metadata.